src/spell/modules/spell/spell/src/auth/mod.rs

```diff
@@ -0,0 +1,185 @@
+mod roles;
+mod keys;
+
+pub use self::roles::{is_by_spell, is_by_creator};
+
+use marine_rs_sdk::CallParameters;
+use crate::auth::keys::parse_permission;
+use crate::auth::roles::{Auth, authenticate};
+
+pub fn guard_kv_write(key: &str)  -> eyre::Result<()> {
+    let cp = marine_rs_sdk::get_call_parameters();
+    match authenticate(&cp) {
+        Auth::Spell => Ok(()),
+        Auth::Other(roles) => {
+            let allowed_roles = parse_permission(key);
```

Comment:
`permission`, `role, `auth`, `right` – which is which?

I feel like there should be only two concepts behind this: Permissions in the Key – `hw_`, and Role of the Caller – `Spell, Worker, Host`.

Reply:
And permission is probably just `HashSet<Role>`

Reply:
So I think it should be something like
```rust
let key_permissions = parse_permissions(key);
let role = get_role(&cp);
if key_permissions.contain(role) {
  // all is good
} else {
 // FORBIDDEN!
}
```

Reply:
I also thought so, but then I got that 
1. the spell is always allowed to write anything
2. a caller can have several roles: it can be both host and worker or even host, worker, and spell, so it's not just allowed_permissions.contain(caller_role), it's an intersection of sets of allowed and existing roles

Reply:
But yeah, I got problems with naming things

Reply:
> a caller can have several roles

I'm not sure that's true. I think the `role` could be only a single one. 

- If it's spell's particle, and particle comes from `call_params.worker_id`  => Role::Spell
- If particle comes from `call_params.worker_id` => Role::Worker
- if particle comes from `call_params.host_id` => Role::Host

I don't see when Caller can have multiple Roles.

Reply:
`parse_permissions` would work like that then:
- "key" => [Role::Spell]
- "w_key" => [Role::Spell, Role::Worker]
- "h_key" => [Role::Spell, Role::Host]
- "hw_key" => [Role::Spell, Role::Worker, Role::Host]

There may be some sense in doing the following. But if possible, I would like to avoid it.
- "w_key" => [Spell, Worker, Host]

Reply:
A decider spells is a host, worker for the spells on the root worker, and can change keys with w_ and h_ prefixes, for example

Reply:
Note that in this suggestion (which how I did it before, lol)
```
    "key" => [Role::Spell]
    "w_key" => [Role::Spell, Role::Worker]
    "h_key" => [Role::Spell, Role::Host]
    "hw_key" => [Role::Spell, Role::Worker, Role::Host]
```
a spell role is in every case. So, it can be considered a given that a spell can change everything

Reply:
> A decider spells is a host and a worker for the spells on the root worker, and can change keys with w_ and h_ prefixes, for example

Decider is `worker` on root worker, and `host` for every other spell. 

Always a single role. I think that's the only needed change besides the names.

Reply:
> Decider is worker on root worker, and host for every other spell.

That would mean that Decider cannot install WorkerSpells on `root worker`. Because WorkerSpell will have `h_app_cid`, and Decider would pass both checks: `particle.init_peer_id == params.worker_id == params.host_id`.

So depending on the _order of checks_, Decider would either end up a `Role::Worker`, or a `Role::Host`.

Maybe we should check for `Role::Host` first, and all will be good?

Reply:
So instead of that:
1. If it's spell's particle, and particle comes from call_params.worker_id => Role::Spell
2. If particle comes from call_params.worker_id => Role::Worker
3. if particle comes from call_params.host_id => Role::Host

Change 1 & 3:
1. if particle comes from call_params.host_id => Role::Host
2. If particle comes from call_params.worker_id => Role::Worker
3. If it's spell's particle, and particle comes from call_params.worker_id => Role::Spell

Reply:
Decided to do the following:

First check if it's a spell, then if it's a host, and finally if it's a worker:

1. If it's spell's particle, and particle comes from call_params.worker_id => Role::Spell
2. if particle comes from call_params.host_id => Role::Host
3. If particle comes from call_params.worker_id => Role::Worker